fix(plugin-form): stashed session is destroyed when a new session starts in the same room

## Summary
Form sessions are persisted as components keyed only by room: the component type is `form_session:{roomId}` (storage.ts), and the runtime stores/looks up components by the natural key (entityId, type). Because the key omits the session id, a given (entity, room) can hold only ONE session component at a time. The stash/restore feature relies on a stashed session coexisting with later activity in the same room, but `FormService.stash()` only flips status to "stashed" without changing the storage key. When the user subsequently starts any new form in the same room, `startSession()` -> `saveSession()` calls `getComponent(entityId, "form_session:{roomId}")`, finds the stashed session's component, and overwrites it via `updateComponent`. The stashed session (and all the user's invested work the effort-based TTL is designed to protect) is silently and permanently lost; `getStashedSessions()` and `restore()` can never find it again. This also means two stashed sessions in one room, or a stashed + active session in one room (which `restore()`'s own guard explicitly anticipates), can never both persist.

## Reproduction
```
Ran a focused vitest against the real storage functions (src/storage.ts) with an in-memory component store mirroring the runtime's getComponent(entityId, type) natural-key semantics (confirmed in packages/core/src/runtime.ts:11841 getComponent -> getComponentsByNaturalKeys([{entityId,type,worldId,sourceEntityId}])).

Steps: saveSession(A, status="stashed", roomId=R); assert getStashedSessions()==["sessionA"] (passes); saveSession(B, status="active", roomId=R); assert getStashedSessions() still ==["sessionA"].

Observed: after saving B, getStashedSessions() returns [] — session A's component was overwritten because both map to type "form_session:R".
  AssertionError: expected [] to deeply equal [ 'sessionA' ]
Expected: stashed session A remains retrievable; active session B is the active one.

Command: ./node_modules/.bin/vitest run --config vitest.config.ts src/stash-repro.test.ts (in plugins/plugin-form). Temp test removed after run; git tree left unchanged.
```

## Proposed fix
Key session components by session id, not just room: use type `form_session:{roomId}:{sessionId}` (or store all room sessions and disambiguate by id). Update storage.ts: saveSession/deleteSession must locate the component by (roomId,sessionId); getActiveSession must scan components with prefix `form_session:{roomId}:` and return the single active/ready one (startSession already enforces one active/ready per room); getSessionById/getStashedSessions/getAllActiveSessions already iterate getComponents and only need the prefix change. Scope is contained to storage.ts (session functions) plus type constants. Test plan (>=10 lines): the repro above; a test that two stashed sessions in one room both persist and both list; a test that restore() succeeds after a new session started in the same room; a test that deleteSession removes only the targeted session and leaves siblings; a test that startSession still rejects a second active session in the same room.

## Evidence plan
The PR will attach the failing-before/passing-after vitest (the repro plus the multi-session persistence tests) run with `bun run --cwd plugins/plugin-form test`, and a short before/after transcript of stash -> start-new-form -> restore showing the stashed session recovered instead of lost. Package typecheck/lint and root `bun run verify` output included.

## Duplicate search
gh search issues/prs on elizaOS/eliza (read-only). Queries: "plugin-form", "form session stash", "form stashed session", "form session room", "plugin-form session", "form stash restore", "form multiple sessions room", "form_session roomId component". Found only unrelated/prior plugin-form fixes (#22034 buildTemplateValues overwrite, #21999 number validation) and #20159 (core queryEntities filter). No open or closed report of the per-room session-key collision / stash data-loss.

_Filed under a standing operator authorization for the autonomous audit pipeline, after local reproduction and duplicate search._

## Contribution provenance

- AI assistance: yes
- AI provider/model: anthropic / claude-opus-4-8
- Client / agent tooling: pi coding agent (eliza-swarm, autonomous)
- Attribution status: self-reported

<!-- eliza-computer-attribution:v1 {"provider":"anthropic","model":"claude-opus-4-8","client":"pi-coding-agent/eliza-swarm","skill_revision":"elizaOS/eliza@87248f5b552a876ed8e3f9187e579247a8562067:packages/skills/skills/contribute-to-eliza"} -->

--- a/plugins/plugin-form/AGENTS.md
+++ b/plugins/plugin-form/AGENTS.md
@@ -105,7 +105,7 @@ await formService.startSession('onboard', entityId, roomId, { context: { tier: '
 
 ## Conventions / gotchas
 
-- **Storage uses elizaOS Components** (no custom DB tables). Sessions are keyed `form_session:{roomId}`, submissions as `form_submission:{formId}:{submissionId}`. All CRUD is in `storage.ts`.
+- **Storage uses elizaOS Components** (no custom DB tables). Sessions are keyed `form_session:{roomId}:{sessionId}`, submissions as `form_submission:{formId}:{submissionId}`. All CRUD is in `storage.ts`. The session id is part of the natural key so a stashed session survives a new session started in the same room (multiple stashed sessions can coexist).
 - **One active session per user per room.** Calling `startSession` when one already exists throws. Stash or cancel the existing one first.
 - **`restore` intent must go through the FORM action**, not the evaluator, so the provider has fresh context before the agent reply.
 - **Effort-based TTL.** Sessions don't expire at a fixed time; more user interaction extends retention (min 14 days, max 90 days, configurable per form via `FormDefinition.ttl`).

--- a/plugins/plugin-form/CLAUDE.md
+++ b/plugins/plugin-form/CLAUDE.md
@@ -105,7 +105,7 @@ await formService.startSession('onboard', entityId, roomId, { context: { tier: '
 
 ## Conventions / gotchas
 
-- **Storage uses elizaOS Components** (no custom DB tables). Sessions are keyed `form_session:{roomId}`, submissions as `form_submission:{formId}:{submissionId}`. All CRUD is in `storage.ts`.
+- **Storage uses elizaOS Components** (no custom DB tables). Sessions are keyed `form_session:{roomId}:{sessionId}`, submissions as `form_submission:{formId}:{submissionId}`. All CRUD is in `storage.ts`. The session id is part of the natural key so a stashed session survives a new session started in the same room (multiple stashed sessions can coexist).
 - **One active session per user per room.** Calling `startSession` when one already exists throws. Stash or cancel the existing one first.
 - **`restore` intent must go through the FORM action**, not the evaluator, so the provider has fresh context before the agent reply.
 - **Effort-based TTL.** Sessions don't expire at a fixed time; more user interaction extends retention (min 14 days, max 90 days, configurable per form via `FormDefinition.ttl`).

--- a/plugins/plugin-form/src/service-stash-restore.test.ts
+++ b/plugins/plugin-form/src/service-stash-restore.test.ts
@@ -1,0 +1,106 @@
+/**
+ * Regression test for FormService stash/restore across a same-room restart
+ * (issue #22272). Exercises the real FormService against an in-memory component
+ * store that mirrors the runtime's natural-key (entityId, type) semantics, so
+ * the stash -> start-new-form -> restore path is proven end to end. Deterministic,
+ * no live model.
+ */
+import type { Component, IAgentRuntime, UUID } from "@elizaos/core";
+import { beforeEach, describe, expect, it } from "vitest";
+import { FormService } from "./service";
+import type { FormDefinition } from "./types";
+
+const entityId = "00000000-0000-4000-8000-000000000401" as UUID;
+const roomId = "00000000-0000-4000-8000-000000000402" as UUID;
+const agentId = "00000000-0000-4000-8000-000000000403" as UUID;
+
+function makeRuntime(): IAgentRuntime {
+  const components = new Map<string, Component>();
+  const keyFor = (entity: UUID, type: string) => `${entity}:${type}`;
+
+  return {
+    agentId,
+    getRoom: async () => ({ id: roomId, worldId: agentId }),
+    getComponent: async (entity: UUID, type: string) =>
+      components.get(keyFor(entity, type)),
+    getComponents: async (entity: UUID) =>
+      Array.from(components.values()).filter((c) => c.entityId === entity),
+    createComponent: async (component: Component) => {
+      components.set(keyFor(component.entityId, component.type), component);
+      return true;
+    },
+    updateComponent: async (component: Component) => {
+      components.set(keyFor(component.entityId, component.type), component);
+    },
+    deleteComponent: async (id: UUID) => {
+      for (const [key, component] of components) {
+        if (component.id === id) components.delete(key);
+      }
+    },
+    emitEvent: async () => undefined,
+    registerTaskWorker: () => undefined,
+    getTaskWorker: () => undefined,
+    logger: {
+      debug: () => undefined,
+      error: () => undefined,
+      warn: () => undefined,
+      info: () => undefined,
+    },
+  } as unknown as IAgentRuntime;
+}
+
+function signupForm(): FormDefinition {
+  return {
+    id: "signup",
+    name: "Signup",
+    controls: [
+      { key: "name", label: "Name", type: "text", required: true },
+      { key: "email", label: "Email", type: "email", required: true },
+    ],
+  };
+}
+
+describe("FormService stash/restore across same-room restart (issue #22272)", () => {
+  let service: FormService;
+
+  beforeEach(async () => {
+    service = (await FormService.start(makeRuntime())) as FormService;
+    service.registerForm(signupForm());
+  });
+
+  it("restores a stashed session after a new form starts in the same room", async () => {
+    const first = await service.startSession("signup", entityId, roomId);
+    await service.stash(first.id, entityId);
+
+    // Start a brand-new form in the SAME room. Before the fix this overwrote the
+    // stashed session's component (both keyed form_session:{roomId}), silently
+    // destroying the user's stashed work.
+    const second = await service.startSession("signup", entityId, roomId);
+    expect(second.id).not.toBe(first.id);
+
+    // Stashed session must still be retrievable.
+    const stashed = await service.getStashedSessions(entityId);
+    expect(stashed.map((s) => s.id)).toEqual([first.id]);
+
+    // Clear the room so restore's active-session guard is satisfied.
+    await service.cancel(second.id, entityId, true);
+
+    const restored = await service.restore(first.id, entityId);
+    expect(restored.id).toBe(first.id);
+    expect(restored.status).toBe("active");
+
+    const active = await service.getActiveSession(entityId, roomId);
+    expect(active?.id).toBe(first.id);
+  });
+
+  it("still rejects a second active session in the same room", async () => {
+    const first = await service.startSession("signup", entityId, roomId);
+    expect(first.id).toBeTruthy();
+
+    // A second active session in the same room must be rejected while one is
+    // already active/ready. The keying fix must not weaken this invariant.
+    await expect(
+      service.startSession("signup", entityId, roomId),
+    ).rejects.toThrow(/Active session already exists/);
+  });
+});

--- a/plugins/plugin-form/src/session-key.test.ts
+++ b/plugins/plugin-form/src/session-key.test.ts
@@ -15,7 +15,11 @@ import {
   getStashedSessions,
   saveSession,
 } from "./storage";
-import type { FormSession } from "./types";
+import { FORM_SESSION_COMPONENT, type FormSession } from "./types";
+
+const legacyType = (room: UUID) => `${FORM_SESSION_COMPONENT}:${room}`;
+const sessionType = (room: UUID, id: string) =>
+  `${FORM_SESSION_COMPONENT}:${room}:${id}`;
 
 const agentId = "00000000-0000-4000-8000-000000000301" as UUID;
 const entityId = "00000000-0000-4000-8000-000000000302" as UUID;
@@ -155,6 +159,95 @@ describe("form session component keying (issue #22272)", () => {
     expect((await getActiveSession(runtime, entityId, otherRoomId))?.id).toBe(
       "there",
     );
+  });
+
+  it("retires the pre-upgrade room-only component when stashing after upgrade", async () => {
+    const runtime = makeRuntime();
+
+    // Seed a live session under the legacy room-only key, exactly as the
+    // pre-upgrade code persisted in-flight sessions.
+    const legacy = makeSession("legacy", { status: "active" });
+    await runtime.createComponent({
+      id: "00000000-0000-4000-8000-0000000003a1" as UUID,
+      entityId,
+      agentId,
+      roomId,
+      worldId: agentId,
+      sourceEntityId: agentId,
+      type: legacyType(roomId),
+      createdAt: NOW - 10_000,
+      data: JSON.parse(JSON.stringify(legacy)),
+    } as Component);
+
+    // Stash the in-flight session (loaded by id, re-persisted under new key).
+    await saveSession(runtime, { ...legacy, status: "stashed" });
+
+    // The legacy room-only row is gone; the session lives only under the
+    // session-id key.
+    expect(
+      await runtime.getComponent(entityId, legacyType(roomId)),
+    ).toBeFalsy();
+    const components = await runtime.getComponents(entityId);
+    expect(components.map((c) => c.type)).toEqual([
+      sessionType(roomId, "legacy"),
+    ]);
+
+    // No ghost active session blocks a fresh start in the same room.
+    expect(await getActiveSession(runtime, entityId, roomId)).toBeNull();
+    await saveSession(runtime, makeSession("fresh", { status: "active" }));
+    expect((await getActiveSession(runtime, entityId, roomId))?.id).toBe(
+      "fresh",
+    );
+  });
+
+  it("retires the legacy room-only component on delete", async () => {
+    const runtime = makeRuntime();
+
+    const legacy = makeSession("legacy", { status: "active" });
+    await runtime.createComponent({
+      id: "00000000-0000-4000-8000-0000000003a2" as UUID,
+      entityId,
+      agentId,
+      roomId,
+      worldId: agentId,
+      sourceEntityId: agentId,
+      type: legacyType(roomId),
+      createdAt: NOW - 10_000,
+      data: JSON.parse(JSON.stringify(legacy)),
+    } as Component);
+
+    await deleteSession(runtime, legacy);
+
+    expect(
+      await runtime.getComponent(entityId, legacyType(roomId)),
+    ).toBeFalsy();
+    expect(await getActiveSession(runtime, entityId, roomId)).toBeNull();
+  });
+
+  it("never retires a legacy room-only row owned by a different session", async () => {
+    const runtime = makeRuntime();
+
+    // A legacy row for a DIFFERENT session id must survive when we save/delete
+    // an unrelated session in the same room.
+    const other = makeSession("other", { status: "active" });
+    await runtime.createComponent({
+      id: "00000000-0000-4000-8000-0000000003a3" as UUID,
+      entityId,
+      agentId,
+      roomId,
+      worldId: agentId,
+      sourceEntityId: agentId,
+      type: legacyType(roomId),
+      createdAt: NOW - 10_000,
+      data: JSON.parse(JSON.stringify(other)),
+    } as Component);
+
+    await saveSession(runtime, makeSession("mine", { status: "stashed" }));
+
+    const legacyRow = await runtime.getComponent(entityId, legacyType(roomId));
+    if (!legacyRow)
+      throw new Error("legacy row for a different session was retired");
+    expect((legacyRow.data as FormSession).id).toBe("other");
   });
 
   it("updates a session in place without creating a duplicate component", async () => {

--- a/plugins/plugin-form/src/session-key.test.ts
+++ b/plugins/plugin-form/src/session-key.test.ts
@@ -1,0 +1,176 @@
+/**
+ * Regression tests for session component keying (issue #22272). Sessions must be
+ * keyed by (roomId, sessionId) so a stashed session survives a new session
+ * started in the same room. Runs against an in-memory component store that
+ * mirrors the runtime's natural-key getComponent(entityId, type) semantics
+ * (see packages/core/src/runtime.ts getComponent -> getComponentsByNaturalKeys).
+ * Deterministic, no live model.
+ */
+import type { Component, IAgentRuntime, UUID } from "@elizaos/core";
+import { describe, expect, it } from "vitest";
+import {
+  deleteSession,
+  getActiveSession,
+  getSessionById,
+  getStashedSessions,
+  saveSession,
+} from "./storage";
+import type { FormSession } from "./types";
+
+const agentId = "00000000-0000-4000-8000-000000000301" as UUID;
+const entityId = "00000000-0000-4000-8000-000000000302" as UUID;
+const roomId = "00000000-0000-4000-8000-000000000303" as UUID;
+const otherRoomId = "00000000-0000-4000-8000-000000000304" as UUID;
+// Base timestamps on the real clock so isExpired()/isLiveSession() treat every
+// fixture session as live regardless of when the suite runs.
+const NOW = Date.now();
+
+function makeSession(
+  id: string,
+  overrides: Partial<FormSession> = {},
+): FormSession {
+  return {
+    id,
+    formId: "signup",
+    formVersion: 1,
+    entityId,
+    roomId,
+    status: "active",
+    fields: {},
+    history: [],
+    effort: {
+      interactionCount: 1,
+      timeSpentMs: 1000,
+      firstInteractionAt: NOW - 10_000,
+      lastInteractionAt: NOW - 10_000,
+    },
+    // Well in the future so isLiveSession() is always true.
+    expiresAt: NOW + 86_400_000_000,
+    createdAt: NOW - 10_000,
+    updatedAt: NOW - 10_000,
+    ...overrides,
+  };
+}
+
+/**
+ * In-memory component store keyed exactly like the runtime: by the natural key
+ * (entityId, type). This is the semantics that made the room-only key destroy
+ * stashed sessions.
+ */
+function makeRuntime(): IAgentRuntime {
+  const components = new Map<string, Component>();
+  const keyFor = (entity: UUID, type: string) => `${entity}:${type}`;
+
+  return {
+    agentId,
+    getRoom: async () => ({ id: roomId, worldId: agentId }),
+    getComponent: async (entity: UUID, type: string) =>
+      components.get(keyFor(entity, type)),
+    getComponents: async (entity: UUID) =>
+      Array.from(components.values()).filter((c) => c.entityId === entity),
+    createComponent: async (component: Component) => {
+      components.set(keyFor(component.entityId, component.type), component);
+      return true;
+    },
+    updateComponent: async (component: Component) => {
+      components.set(keyFor(component.entityId, component.type), component);
+    },
+    deleteComponent: async (id: UUID) => {
+      for (const [key, component] of components) {
+        if (component.id === id) components.delete(key);
+      }
+    },
+  } as unknown as IAgentRuntime;
+}
+
+describe("form session component keying (issue #22272)", () => {
+  it("keeps a stashed session when a new session starts in the same room", async () => {
+    const runtime = makeRuntime();
+
+    await saveSession(runtime, makeSession("sessionA", { status: "stashed" }));
+    expect(
+      (await getStashedSessions(runtime, entityId)).map((s) => s.id),
+    ).toEqual(["sessionA"]);
+
+    // Starting a new active session in the SAME room previously overwrote the
+    // stashed component because both mapped to type form_session:{roomId}.
+    await saveSession(runtime, makeSession("sessionB", { status: "active" }));
+
+    const stashed = await getStashedSessions(runtime, entityId);
+    expect(stashed.map((s) => s.id)).toEqual(["sessionA"]);
+
+    const active = await getActiveSession(runtime, entityId, roomId);
+    expect(active?.id).toBe("sessionB");
+
+    // The stashed session is still individually retrievable by id.
+    expect((await getSessionById(runtime, entityId, "sessionA"))?.id).toBe(
+      "sessionA",
+    );
+  });
+
+  it("persists two stashed sessions in one room and lists both", async () => {
+    const runtime = makeRuntime();
+
+    await saveSession(runtime, makeSession("stashA", { status: "stashed" }));
+    await saveSession(runtime, makeSession("stashB", { status: "stashed" }));
+
+    const stashedIds = (await getStashedSessions(runtime, entityId))
+      .map((s) => s.id)
+      .sort();
+    expect(stashedIds).toEqual(["stashA", "stashB"]);
+  });
+
+  it("deletes only the targeted session and leaves siblings intact", async () => {
+    const runtime = makeRuntime();
+
+    const stashed = makeSession("stashA", { status: "stashed" });
+    const active = makeSession("activeB", { status: "active" });
+    await saveSession(runtime, stashed);
+    await saveSession(runtime, active);
+
+    await deleteSession(runtime, active);
+
+    expect(await getSessionById(runtime, entityId, "activeB")).toBeNull();
+    expect((await getSessionById(runtime, entityId, "stashA"))?.id).toBe(
+      "stashA",
+    );
+    expect(await getActiveSession(runtime, entityId, roomId)).toBeNull();
+  });
+
+  it("scopes active-session lookup to the requested room", async () => {
+    const runtime = makeRuntime();
+
+    await saveSession(
+      runtime,
+      makeSession("here", { status: "active", roomId }),
+    );
+    await saveSession(
+      runtime,
+      makeSession("there", { status: "active", roomId: otherRoomId }),
+    );
+
+    expect((await getActiveSession(runtime, entityId, roomId))?.id).toBe(
+      "here",
+    );
+    expect((await getActiveSession(runtime, entityId, otherRoomId))?.id).toBe(
+      "there",
+    );
+  });
+
+  it("updates a session in place without creating a duplicate component", async () => {
+    const runtime = makeRuntime();
+
+    const session = makeSession("sessionA", { status: "active" });
+    await saveSession(runtime, session);
+    await saveSession(runtime, { ...session, status: "stashed" });
+
+    const all = await getSessionById(runtime, entityId, "sessionA");
+    expect(all?.status).toBe("stashed");
+    // Exactly one active-or-stashed component should exist for this session.
+    const components = await runtime.getComponents(entityId);
+    const sessionComponents = components.filter((c) =>
+      c.type.includes("sessionA"),
+    );
+    expect(sessionComponents).toHaveLength(1);
+  });
+});

--- a/plugins/plugin-form/src/storage.ts
+++ b/plugins/plugin-form/src/storage.ts
@@ -27,6 +27,9 @@
  * - The session id is part of the component type so distinct sessions in the
  *   same room map to distinct natural keys and never overwrite each other
  * - Room scoping ensures different rooms have different contexts
+ * - Deployments upgrading from the pre-session-id `form_session:{roomId}` key
+ *   have their in-flight session's legacy component retired on the next
+ *   saveSession/deleteSession so the stale row cannot shadow the new key
  *
  * ### Submissions
  * - Stored as components with type: `form_submission:{formId}:{submissionId}`
@@ -104,6 +107,44 @@ const isLiveSession = (session: FormSession): boolean => !isExpired(session);
  */
 const sessionComponentType = (roomId: UUID, sessionId: string): string =>
   `${FORM_SESSION_COMPONENT}:${roomId}:${sessionId}`;
+
+/**
+ * Build the pre-upgrade component type that keyed a session by room only.
+ *
+ * WHY this still matters:
+ * - Deployments that ran the old code persisted in-flight sessions under
+ *   `form_session:{roomId}` (one row per room, no session id).
+ * - After upgrading, saveSession/deleteSession write the new session-id key,
+ *   which is a different natural key, so getComponent misses the legacy row.
+ * - Left untouched, that stale `active` row keeps satisfying getActiveSession
+ *   and blocks startSession. retireLegacySessionComponent removes it once the
+ *   owning session has been re-persisted (or deleted) under the new key.
+ */
+const legacySessionComponentType = (roomId: UUID): string =>
+  `${FORM_SESSION_COMPONENT}:${roomId}`;
+
+/**
+ * Delete the pre-upgrade room-only component for a session, if present.
+ *
+ * Only removes the legacy row when it actually holds this session's data, so a
+ * room-only row belonging to a different session id is never touched.
+ */
+const retireLegacySessionComponent = async (
+  runtime: IAgentRuntime,
+  session: FormSession,
+): Promise<void> => {
+  const legacy = await runtime.getComponent(
+    session.entityId,
+    legacySessionComponentType(session.roomId),
+  );
+  if (
+    legacy?.data &&
+    isFormSession(legacy.data) &&
+    legacy.data.id === session.id
+  ) {
+    await runtime.deleteComponent(legacy.id);
+  }
+};
 
 const RESTORABLE_SESSION_STATUSES: FormSession["status"][] = [
   "active",
@@ -370,6 +411,10 @@ export async function saveSession(
   } else {
     await runtime.createComponent(component);
   }
+
+  // Retire any pre-upgrade room-only component for this session so the stale
+  // `active` row cannot shadow the freshly written session-id key.
+  await retireLegacySessionComponent(runtime, session);
 }
 
 /**
@@ -393,6 +438,10 @@ export async function deleteSession(
   if (existing) {
     await runtime.deleteComponent(existing.id);
   }
+
+  // Also remove a pre-upgrade room-only component so deleting a session leaves
+  // no ghost row behind under the legacy key.
+  await retireLegacySessionComponent(runtime, session);
 }
 
 // ============================================================================

--- a/plugins/plugin-form/src/storage.ts
+++ b/plugins/plugin-form/src/storage.ts
@@ -21,9 +21,12 @@
  * ## Storage Strategy
  *
  * ### Sessions
- * - Stored as components with type: `form_session:{roomId}`
- * - One active session per user per room
- * - Scoping ensures different rooms have different contexts
+ * - Stored as components with type: `form_session:{roomId}:{sessionId}`
+ * - One active/ready session per user per room, but multiple stashed
+ *   sessions (and a stashed session alongside a new active one) can coexist
+ * - The session id is part of the component type so distinct sessions in the
+ *   same room map to distinct natural keys and never overwrite each other
+ * - Room scoping ensures different rooms have different contexts
  *
  * ### Submissions
  * - Stored as components with type: `form_submission:{formId}:{submissionId}`
@@ -87,6 +90,20 @@ const isFormSession = (data: JsonValue | object): data is FormSession => {
 };
 
 const isLiveSession = (session: FormSession): boolean => !isExpired(session);
+
+/**
+ * Build the component type (natural key suffix) for a session.
+ *
+ * WHY the session id is included:
+ * - The runtime stores/looks up components by the natural key (entityId, type).
+ * - Keying only by room (`form_session:{roomId}`) let a single (entity, room)
+ *   hold just one session component, so stashing a session and then starting a
+ *   new one in the same room silently overwrote and destroyed the stash.
+ * - Including the session id gives every session its own natural key, so a
+ *   stashed session survives new activity in the same room.
+ */
+const sessionComponentType = (roomId: UUID, sessionId: string): string =>
+  `${FORM_SESSION_COMPONENT}:${roomId}:${sessionId}`;
 
 const RESTORABLE_SESSION_STATUSES: FormSession["status"][] = [
   "active",
@@ -182,23 +199,27 @@ export async function getActiveSession(
   entityId: UUID,
   roomId: UUID,
 ): Promise<FormSession | null> {
-  // Component type includes roomId for room-level scoping
-  const component = await runtime.getComponent(
-    entityId,
-    `${FORM_SESSION_COMPONENT}:${roomId}`,
-  );
+  // Session components are now keyed by (roomId, sessionId), so we scan the
+  // entity's session components and match on the session's own roomId rather
+  // than reading a single room-keyed component. startSession enforces at most
+  // one active/ready session per room, so the first match is the active one.
+  const components = await runtime.getComponents(entityId);
 
-  if (!component?.data || !isFormSession(component.data)) return null;
+  for (const component of components) {
+    if (!component.type.startsWith(`${FORM_SESSION_COMPONENT}:`)) continue;
+    if (!component.data || !isFormSession(component.data)) continue;
 
-  const session = component.data;
+    const session = component.data;
 
-  // Only return if active (not stashed, submitted, cancelled, or expired)
-  // WHY: Other statuses require explicit action to restore/continue
-  if (
-    isLiveSession(session) &&
-    (session.status === "active" || session.status === "ready")
-  ) {
-    return session;
+    // Only return if active (not stashed, submitted, cancelled, or expired)
+    // WHY: Other statuses require explicit action to restore/continue
+    if (
+      session.roomId === roomId &&
+      isLiveSession(session) &&
+      (session.status === "active" || session.status === "ready")
+    ) {
+      return session;
+    }
   }
 
   return null;
@@ -325,7 +346,7 @@ export async function saveSession(
   runtime: IAgentRuntime,
   session: FormSession,
 ): Promise<void> {
-  const componentType = `${FORM_SESSION_COMPONENT}:${session.roomId}`;
+  const componentType = sessionComponentType(session.roomId, session.id);
   const existing = await runtime.getComponent(session.entityId, componentType);
   const context = await resolveComponentContext(runtime, session.roomId);
   const resolvedWorldId = existing?.worldId ?? context.worldId;
@@ -366,7 +387,7 @@ export async function deleteSession(
   runtime: IAgentRuntime,
   session: FormSession,
 ): Promise<void> {
-  const componentType = `${FORM_SESSION_COMPONENT}:${session.roomId}`;
+  const componentType = sessionComponentType(session.roomId, session.id);
   const existing = await runtime.getComponent(session.entityId, componentType);
 
   if (existing) {

--- a/plugins/plugin-form/src/types.ts
+++ b/plugins/plugin-form/src/types.ts
@@ -1429,6 +1429,11 @@ export const FORM_DEFINITION_DEFAULTS = {
  * - Components are elizaOS's entity data storage
  * - Scoped to entity, can include room in type
  * - Automatic CRUD via runtime
+ *
+ * Full session component type is `form_session:{roomId}:{sessionId}`. The
+ * session id is part of the natural key so multiple sessions in one room
+ * (e.g. a stashed session plus a newly started one) never overwrite each
+ * other in the (entityId, type) component store.
  */
 export const FORM_SESSION_COMPONENT = "form_session";
 

